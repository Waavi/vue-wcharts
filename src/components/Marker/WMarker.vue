
<template>
    <g>
        <line
            v-bind="line"
            :stroke="stylesCmp.stroke"
            :stroke-width="stylesCmp.strokeWidth"
            :stroke-dasharray="stylesCmp.strokeDasharray"
            fill="none"
        />
        <slot
            v-bind="labelCoordinates"
            :isX="isX"
            :label="label"
            :value="value"
            :placement="placement"
            :style="labelStylesCmp"
        >
            <text
                v-bind="labelCoordinates"
                :text-anchor="isX ? 'end' : placement"
                :style="labelStylesCmp"
            >
                {{ label }}
            </text>
        </slot>
    </g>
</template>

<script>
import VueTypes from 'vue-types'

const stylesDefaultProp = {
    stroke: '#333',
    strokeWidth: 1,
    strokeDasharray: '0',
}
const labelStylesDefaultProp = {
    fill: '#333',
    fontSize: 14,
}

export default {
    name: 'WMarker',
    type: 'cartesian',
    inject: ['Chart'],
    props: {
        type: VueTypes.oneOf(['y', 'x']).def('y'),
        value: VueTypes.any.isRequired,
        label: VueTypes.string.def(''),
        placement: VueTypes.oneOf(['start', 'end']).def('end'),
        styles: VueTypes.shape({
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            strokeDasharray: VueTypes.oneOfType([VueTypes.string, VueTypes.number]),
        }).def(() => ({
            ...stylesDefaultProp,
        })),
        labelStyles: VueTypes.object.def(() => ({
            ...labelStylesDefaultProp,
        })),
    },
    computed: {
        // Generate styles of line
        stylesCmp () {
            return {
                ...stylesDefaultProp,
                ...this.styles,
            }
        },
        // Generate styles of label
        labelStylesCmp () {
            return {
                ...labelStylesDefaultProp,
                ...this.labelStyles,
            }
        },
        isX () {
            return this.type === 'x'
        },
        line () {
            const {
                dataset, canvas, yScale, padding, axis,
            } = this.Chart
            if (this.isX) {
                // Calculate left and right space
                const offset = padding[1] + padding[3]
                // Calculate space between lines
                const space = (canvas.width - offset) / (dataset.length - 1)
                const index = dataset.findIndex(d => d[axis.x.datakey] === this.value)
                const x = canvas.x0 + (space * index) + padding[3]
                return {
                    x1: x,
                    y1: canvas.y0,
                    x2: x,
                    y2: canvas.y1,
                }
            }
            const y = yScale(this.value)
            return {
                x1: canvas.x0,
                y1: y,
                x2: canvas.x1,
                y2: y,
            }
        },
        labelCoordinates () {
            if (this.isX) {
                return {
                    x: this.line.x1 - 5,
                    y: this.placement === 'end' ? this.line.y1 : this.line.y2 + 5,
                    dy: this.placement === 'end' ? 15 : -10,
                }
            }
            return {
                x: this.placement === 'end' ? this.line.x2 - 5 : this.line.x1 + 5,
                y: this.line.y1,
                dy: -10,
            }
        },
    },
}
</script>
