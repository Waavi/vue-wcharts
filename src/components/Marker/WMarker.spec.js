import { shallowMount } from '@vue/test-utils'
import WMarker from './WMarker.vue'

describe('Components/WMarker', () => {
    const dataset = [
        {
            name: 'Page A', one: 4000, two: -2400, three: 2300, four: 1200, five: 2300,
        },
        {
            name: 'Page B', one: 3000, two: 2210, three: 1398, four: 1200, five: 1398,
        },
        {
            name: 'Page C', one: 9800, two: 2200, three: 0, four: 1200, five: 0,
        },
    ]

    const provide = {
        Chart: {
            dataset,
            data: dataset,
            axis: { x: { datakey: 'name', name: '' } },
            legends: ['One Bar', 'Two Bar'],
            space: [20, 20, 20, 20],
            padding: [20, 20, 20, 20],
            activeElements: [0, 1],
            yScale: a => a,
            canvas: {
                x0: 40,
                y0: 10,
                width: 540,
                height: 366,
                x1: 580,
                y1: 376,
            },
        },
    }

    const defaultConfigForProps = (propsData = {}) => ({
        propsData,
        provide,
    })

    function snapshot (description, alias, props) {
        it(`render a "${description}" ("${alias}") correctly`, () => {
            const wrapper = shallowMount(WMarker, defaultConfigForProps(props))
            expect(wrapper.html()).toMatchSnapshot()
        })
    }

    snapshot('vertical marker', 'xLine', { x: 10 })
    snapshot('horizontal marker', 'yLine', { y: 10 })
    snapshot('point marker', 'point', { x: 10, y: 20 })
    snapshot('vertical range marker', 'xRange', { x: [10, 20] })
    snapshot('horizontal range marker', 'yRange', { y: [10, 20] })
    snapshot('area marker', 'rect', { x: [10, 20], y: [30, 40] })
})
