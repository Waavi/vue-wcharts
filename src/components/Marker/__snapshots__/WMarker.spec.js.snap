// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WMarker Should be render correctly 1`] = `
<g>
  <line stroke="#333333" stroke-width="1" stroke-dasharray="0" fill="none" x1="40" y1="[object Object]" x2="580" y2="[object Object]"></line> <text text-anchor="end" x="575" y="[object Object]" dy="-10" style="fill: #333333; font-size: 14;">

  </text>
</g>
`;
